File: packages/node/src/resources/projects.ts
import type { MissionInbox } from '../client.js';
import type {
  AssignDomainsToProjectParams,
  AssignDomainsToProjectResult,
  CreateProjectParams,
  Project,
  UpdateProjectParams,
} from '../types.js';

/**
 * The `projects` resource. Access via `mi.projects`.
 *
 * Projects group domains for organisation and reporting. A domain belongs to
 * at most one project.
 */
export class Projects {
  /** @internal */
  constructor(private readonly client: MissionInbox) {}

  /** List all projects on the account. */
  async list(): Promise<Project[]> {
    return this.client.request<Project[]>({
      method: 'GET',
      path: '/api/projects',
    });
  }

  /** Retrieve a project by id. */
  async get(id: string): Promise<Project> {
    return this.client.request<Project>({
      method: 'GET',
      path: `/api/projects/${encodeURIComponent(id)}`,
    });
  }

  /** Create a new project. */
  async create(params: CreateProjectParams): Promise<Project> {
    return this.client.request<Project>({
      method: 'POST',
      path: '/api/projects',
      body: params,
    });
  }

  /** Update a project's name. */
  async update(id: string, params: UpdateProjectParams): Promise<Project> {
    return this.client.request<Project>({
      method: 'PATCH',
      path: `/api/projects/${encodeURIComponent(id)}`,
      body: params,
    });
  }

  /** Delete a project. Domains previously assigned become unassigned. */
  async delete(id: string): Promise<{ message: string }> {
    return this.client.request<{ message: string }>({
      method: 'DELETE',
      path: `/api/projects/${encodeURIComponent(id)}`,
    });
  }

  /**
   * Move domains into a project. Domains previously in another project are
   * reassigned and the response reports the previous owner per domain.
   */
  async assignDomains(id: string, params: AssignDomainsToProjectParams): Promise<AssignDomainsToProjectResult> {
    return this.client.request<AssignDomainsToProjectResult>({
      method: 'PATCH',
      path: `/api/projects/${encodeURIComponent(id)}/domains`,
      body: params,
    });
  }
}

File: packages/python/src/missioninbox/resources/projects.py
"""The `projects` resource. Access via `mi.projects`."""

from __future__ import annotations

from typing import Any, Optional
from urllib.parse import quote

from .._http import HttpClient


class Projects:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(self) -> list[dict[str, Any]]:
        return self._http.request("GET", "/api/projects")

    def get(self, id: str) -> dict[str, Any]:
        return self._http.request("GET", f"/api/projects/{quote(id, safe='')}")

    def create(self, *, name: str) -> dict[str, Any]:
        return self._http.request("POST", "/api/projects", body={"name": name})

    def update(self, id: str, *, name: Optional[str] = None) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        return self._http.request("PATCH", f"/api/projects/{quote(id, safe='')}", body=body)

    def delete(self, id: str) -> dict[str, Any]:
        return self._http.request("DELETE", f"/api/projects/{quote(id, safe='')}")

    def assign_domains(self, id: str, *, domain_names: list[str]) -> dict[str, Any]:
        """Move domains into this project."""
        return self._http.request(
            "PATCH",
            f"/api/projects/{quote(id, safe='')}/domains",
            body={"domainNames": domain_names},
        )
